tests/integrations/wandb/workspace_test.py

```diff
@@ -0,0 +1,160 @@
+        for wandb_run in self.workspace.wandb_client.runs(
+            f"{WANDB_ENTITY}/{WANDB_PROJECT}",
+        ):
+            if (
+                self.UNIQUE_ID_SUFFIX in wandb_run.id
+                or self.UNIQUE_ID_SUFFIX in wandb_run.config.get("_run_suite_id", "")
+            ):
+                wandb_run.delete(delete_artifacts=True)
+
+        teardown_logging()
+
+    def test_direct_usage(self):
+        params = Params.from_file(self.FIXTURES_ROOT / "experiment" / "hello_world.jsonnet")
+        step_graph = StepGraph.from_params(params.pop("steps", keep_as_dict=True))
+        tango_run = self.workspace.register_run(step for step in step_graph.values())
+
+        # Test 'registered_run()' and 'registered_runs()' methods.
+        assert self.workspace.registered_run(tango_run.name) == tango_run
+        assert self.workspace.registered_runs()[tango_run.name] == tango_run
+
+        hello_step = step_graph["hello"]
+        hello_world_step = step_graph["hello_world"]
+
+        # Test getting step info.
+        step_info = self.workspace.step_info(hello_step)
+        assert step_info.unique_id.endswith(self.UNIQUE_ID_SUFFIX)
+        assert step_info.step_name == "hello"
+        assert step_info.state == StepState.INCOMPLETE
+
+        # Mark the "hello" step as starting.
+        self.workspace.step_starting(hello_step)
+        assert self.workspace.step_info(hello_step).state == StepState.RUNNING
+
+        # Mark the "hello" step as finished.
+        self.workspace.step_finished(hello_step, "hello")
+        assert self.workspace.step_info(hello_step).state == StepState.COMPLETED
+
+        # Make sure the result is in the cache, exists locally, and on W&B.
+        cache = self.workspace.cache
+        assert hello_step in cache
+        assert cache.step_dir(hello_step).is_dir()
+        assert cache.get_step_result_artifact(hello_step) is not None
+
+        # Now make sure we can fetch the item from the cache, even if it's not in memory
+        # or in the cache directory.
+        if hello_step.unique_id in cache.weak_cache:
+            del cache.weak_cache[hello_step.unique_id]
+        if hello_step.unique_id in cache.strong_cache:
+            del cache.strong_cache[hello_step.unique_id]
+        shutil.rmtree(cache.step_dir(hello_step))
+        assert hello_step in cache
+        assert cache[hello_step] == "hello"
+
+        # Now start the "hello_world" step and then mark it as failed.
+        self.workspace.step_starting(hello_world_step)
+        self.workspace.step_failed(hello_world_step, ValueError("oh no!"))
+        assert self.workspace.step_info(hello_world_step).state == StepState.FAILED
+
+    @pytest.mark.parametrize(
+        "distributed",
+        [
+            pytest.param(True, id="distributed"),
+            pytest.param(False, id="single-device"),
+        ],
+    )
+    def test_with_wandb_train_callback(self, distributed: bool):
+        self.run(
```

Comment:
We should also have a test with `multicore = True` in `run()`.

Reply:
Adding that to the TODO list 👍

Reply:
Done: 143f23c